scripts/preprocess_data.py

```diff
@@ -0,0 +1,192 @@
+        if line.startswith('set YEAR'):
+            start_year = line.split(' ')[3]
+        if line.startswith('set COMMODITY'): # Extracts list of COMMODITIES from data file. Some models use FUEL instead. 
+            fuel_list = line.split(' ')[3:-1]
+        if line.startswith('set FUEL'): # Extracts list of FUELS from data file. Some models use COMMODITIES instead. 
+            fuel_list = line.split(' ')[3:-1]
+        if line.startswith('set TECHNOLOGY'):
+            tech_list = line.split(' ')[3:-1]
+        if line.startswith('set STORAGE'):
+            storage_list = line.split(' ')[3:-1]
+        if line.startswith('set MODE_OF_OPERATION'):
+            mode_list = line.split(' ')[3:-1]
+
+with open(data_infile, 'r') as f:
+    for line in f:
+        if line.startswith(";"):
+            parsing = False   
+        if parsing:
+            if line.startswith('['):
+                fuel = line.split(',')[2]
+                tech = line.split(',')[1]
+            elif line.startswith(start_year):
+                years = line.rstrip().split(' ')[0:]
+                years = [i.strip(':=') for i in years]
+            elif not line.startswith(start_year):
+                values = line.rstrip().split(' ')[1:]
+                mode = line.split(' ')[0]
+                data_out.append(tuple([fuel,tech,mode]))
+                data_all.append(tuple([tech,mode]))
+                for i in range(0,len(years)):
+                    output_table.append(tuple([tech,fuel,mode,years[i],values[i]]))
+        if line.startswith('param OutputActivityRatio'):
+            parsing = True
+
+with open(data_infile, 'r') as f:
+    for line in f:
+        if line.startswith(";"):
+            parsing = False   
+        if parsing:
+            if line.startswith('['):
+                fuel = line.split(',')[2]
+                tech = line.split(',')[1]
+            elif not line.startswith(start_year):
+                mode = line.split(' ')[0]
+                data_inp.append(tuple([fuel,tech,mode]))
+                data_all.append(tuple([tech,mode]))
+        if line.startswith('param InputActivityRatio'):
+            parsing = True
+
+# For TechnologyToStorage and TechnologyFromStorage 
+storage_to = []
+storage_from = []
+
+with open(data_infile) as f:
+    for line in f:
+        if line.startswith(";"):
+            parsing = False   
+        if parsing:
+            if line.startswith('['):
+                tech = line.split(',')[1]
+            elif not line.startswith(mode_list[0]):
+                storage = line.split(' ')[0]
+                values = line.rstrip().split(' ')[1:]
+                for i in range(0,len(mode_list)):
+                    if values[i] != '0':
+                        storage_to.append(tuple([storage,tech,mode_list[i]]))
+                        #data_all.append(tuple([tech,mode_list[i]]))
+        if line.startswith('param TechnologyToStorage'):
+            parsing = True
+
+with open(data_infile) as f:
```

Comment:
Another issue with this approach is that the `data_infile` is read for each parsing block, which results in reading the same file from disk at least six times. It would be more efficient to open the file once, and iterate over the lines once, and read in all the parameter values within that loop.